### PR TITLE
Add support for Python 3.13; drop support for Python 3.8

### LINKED ISSUE #1056: Add support for Python 3.13

### PR DESCRIPTION
**Pull request by chrisrink10**
Fixes #1056 

I ran [`pyupgrade --py39-plus`](https://github.com/asottile/pyupgrade) on the entire source tree to convert from `typing.{List,Set,...}` to the generic forms of the builtins (among other available upgrades).

Upgrading to 3.13 also surfaced #1088 which I filed to address separately since this PR is already quite large.

### CODE REVIEW

**Comment by ikappaki** [comment]
Hi @chrisrink10, apologies for the intrusion, MS-Windows failure below is due to https://github.com/python/cpython/issues/125283 because of change in `os.path.isabs` behavior, in case you were wondering:

```
_________________________________ reader-test _________________________________
ERROR in (reader-test) (test_io.lpy:86)

Traceback (most recent call last):
  File "D:\a\basilisp\basilisp\tests\basilisp\test_io.lpy", line 86, in reader_test
    (is (= "hi there" (slurp (str "file:///" filename))))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\a\basilisp\basilisp\.tox\py313\Lib\site-packages\basilisp\core.lpy", line 7149, in slurp
    (with-open [reader (apply basilisp.io/reader f opts)]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\a\basilisp\basilisp\.tox\py313\Lib\site-packages\basilisp\core.lpy", line 207, in apply
    (basilisp.lang.runtime/apply f args)))
                           ^^^^^^^^^^^^^^
  File "D:\a\basilisp\basilisp\.tox\py313\Lib\site-packages\basilisp\lang\runtime.py", line 1311, in apply
    return f(*final)
  File "D:\a\basilisp\basilisp\.tox\py313\Lib\site-packages\basilisp\io.lpy", line 291, in reader
    (defn reader
    ...<28 lines>...
           (make-reader f)))
  File "C:\hostedtoolcache\windows\Python\3.13.0\x64\Lib\functools.py", line 929, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "D:\a\basilisp\basilisp\.tox\py313\Lib\site-packages\basilisp\io.lpy", line 210, in __make_reader_str_5880
    (try
    ^^^^
    ...<2 lines>...
        (make-reader (pathlib/Path f) opts))))
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.13.0\x64\Lib\functools.py", line 929, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "D:\a\basilisp\basilisp\.tox\py313\Lib\site-packages\basilisp\io.lpy", line 249, in __make_reader_ParseResult_5860
    (-> (as-path f)
        (make-reader opts))
  File "C:\hostedtoolcache\windows\Python\3.13.0\x64\Lib\functools.py", line 929, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "D:\a\basilisp\basilisp\.tox\py313\Lib\site-packages\basilisp\io.lpy", line 233, in __make_reader_Path_5906
      (apply-kw python/open f opts))
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<55 lines>...
                {:url f :opts opts}))))
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\a\basilisp\basilisp\.tox\py313\Lib\site-packages\basilisp\core.lpy", line 828, in apply_kw
    (basilisp.lang.runtime/apply-kw f args))
                           ^^^^^^^^^^^^^^^^^
  File "D:\a\basilisp\basilisp\.tox\py313\Lib\site-packages\basilisp\lang\runtime.py", line 1336, in apply_kw
    return f(*final, **kwargs)
OSError: [Errno 22] Invalid argument: '\\C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpc1675mnk'
```

**Comment by ikappaki** [comment]
Something along the following fixes the break

```diff
diff --git a/src/basilisp/io.lpy b/src/basilisp/io.lpy
index a07e861..9dca382 100644
--- a/src/basilisp/io.lpy
+++ b/src/basilisp/io.lpy
@@ -34,7 +34,7 @@
       (let [path (.-path f)]
         ;; On MS-Windows, extracting an absolute path from the URL
         ;; incorrectly adds a leading `/', .e.g. /C:\xyz.
-        (pathlib/Path #?(:windows (if (os.path/isabs path) (subs path 1) path)
+        (pathlib/Path #?(:windows (if (str/starts-with? path "/") (subs path 1) path)
                          :default path)))
       (throw
        (ex-info "Cannot coerce non-File URL to pathlib.Path"

```

**Comment by chrisrink10** [comment]
@ikappaki thanks! I pushed that and will see how it goes.

File: .github/workflows/run-tests.yml [hunk 1]
```diff
@@ -188,7 +188,7 @@ jobs:
           tox run-parallel -p 2
 
   report-coverage:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-22.04
```
Comment (chrisrink10) [inline .github/workflows/run-tests.yml:191]: https://github.com/basilisp-lang/basilisp/issues/1089

File: src/basilisp/importer.py [hunk 2]
```diff
@@ -412,7 +404,7 @@ def exec_module(self, module: types.ModuleType) -> None:
         else:
             try:
                 self._exec_cached_module(fullname, spec.loader_state, path_stats, ns)
-            except (EOFError, ImportError, IOError, OSError) as e:
+            except (EOFError, ImportError, OSError) as e:
```
Comment (chrisrink10) [inline src/basilisp/importer.py:407]: [TIL `IOError` is just an alias of `OSError`.](https://docs.python.org/3/library/exceptions.html#IOError)

**Pull request merged**